### PR TITLE
nn/utils: skip no-op .to(first_device) in _get_total_norm

### PR DESCRIPTION
**Pull request by notandruu**
Towards #133586. Replaces #184072 (accidentally closed; the branch was force-pushed afterwards so it cannot be reopened, and its history had no common ancestor with main after a shallow-clone push).

## Summary

For N gradient tensors, the final norm aggregation in `_get_total_norm` always runs `[norm.to(first_device) for norm in norms]`, dispatching N Python method calls that do no data movement in the overwhelmingly common case where all gradients live on one device. This gates the comprehension on an O(k) check over the already-computed device groups (k = unique device/dtype groups, typically 1-3).

For 2000 tensors of 512 elements on CPU (the regime from #133586): `get_total_norm` median 2.45ms -> 2.11ms (~14% faster, M3 Max).

## Why the flat-cat path from #184072 is dropped

While replanting the branch I re-validated the second optimization from #184072 (single `vector_norm` over `torch.cat` of small tensors) and found it loses accuracy versus the hierarchical norm-of-per-tensor-norms reduction: for 2000x512 float32 tensors, relative error vs a float64 reference is **3.1e-5 (flat-cat) vs 4.2e-8 (current foreach path)** — a user-visible numerics change in `clip_grad_norm_`, so it's gone. Only the dispatch-overhead fix remains, which is bitwise-neutral.

All review feedback from #184072 is incorporated: tests live in `test/test_nn.py` on `TestNNDeviceType` (`test_clip_grad_norm_skip_to_single_device`). cc @janeyx99

## Test plan

```
python test/test_nn.py -k clip_grad
Ran 36 tests ... OK (skipped=19)
```

Authored with the help of an AI assistant (Claude Code).

### CODE REVIEW

File: test/test_nn.py [hunk 1]
```diff
@@ -14518,6 +14518,16 @@ def compare_scaling(grads):
             self.assertEqual(len(w), 1)
             self.assertEqual(str(w[0].message), "`parameters` is an empty generator, no gradient clipping will occur.")
 
+    # issue #133586: skip no-op .to(first_device) in _get_total_norm
+    @onlyCPU
+    @parametrize_test('norm_type', (1.0, 2.0, float('inf')))
+    def test_clip_grad_norm_skip_to_single_device(self, norm_type, device):
```
Comment (janeyx99) [inline test/test_nn.py]: i would suspect this is already tested in this file somewhere in another test, no?

**Review by janeyx99 — COMMENTED** [review]
looks good to me, please help clarify test addition

**Comment by notandruu** [comment]
Good call @janeyx99 — the aggregation path is indeed already covered by `test_clip_grad_norm` (foreach False/True × 5 norm types, asserted against a manually computed reference on a single device, which exercises exactly the comprehension this PR gates). The only combination missing was `foreach=None` — the default path. Dropped the standalone test and added `None` to the existing parametrize instead, so the whole existing battery now also runs the default path. Net test diff is one line.

**Comment by janeyx99** [comment]
@claude do a review, is this too extra of a change?

**Comment by notandruu** [comment]
Fair question, and the review above is a fair summary. Two notes and then it's your call:

- Trimmed the comment to one line as suggested.
- On "atypical regime": 2000 small tensors is the workload #133586 actually reports — CPU training with many small parameter tensors where `clip_grad_norm_` overhead shows up on every step. So the narrow regime is the regime the issue is about. That said, the absolute win is modest and the original PR's larger half (flat-cat) is gone for good reasons, so I get that what's left is borderline.

If you feel the remaining micro-opt doesn't carry its weight, I'm fine closing this and the issue can note the conclusion — no attachment. If it's worth keeping, it's ready as-is (+1 line of test parametrize, 7 LOC change, bitwise-neutral).

**Comment by notandruu** [comment]
Friendly ping @janeyx99 — happy to close if you think the value isn't there, just want to get a final call so it doesn't sit open indefinitely.

**Review by janeyx99 — COMMENTED** [review]
@notandruu sorry for the delay, can you run some more benchmarks across (50, 200, 500, 2000) params? Here it doesn't matter their size, right?

Also I'd appreciate if you concised up the AI-sounding comments a little bit! Or summarize your own thoughts in addition.
